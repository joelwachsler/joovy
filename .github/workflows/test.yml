name: Test
on:
  push:
    tags:
      - '!**'
jobs:
  test:
    runs-on: self-hosted
    steps:
      - name: Setup build tools
        run: |
          sudo apt-get update
          sudo apt-get install build-essential autoconf automake g++ libtool -y

      - name: Checkout
        uses: actions/checkout@v2

      - name: Setup node
        uses: actions/setup-node@v2
        with:
          node-version: '16'

      - name: Setup yarn
        run: npm install -g yarn

      - name: Setup cache
        uses: actions/cache@v2
        id: yarn-cache
        with:
          path: '**/node_modules'
          key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}

      - name: Install dependencies
        if: steps.yarn-cache.outputs.cache-hit != 'true'
        run: yarn

      - name: Run tests
        run: yarn test

  lint:
    runs-on: self-hosted
    steps:
      - name: Setup build tools
        run: |
          sudo apt-get update
          sudo apt-get install build-essential autoconf automake g++ libtool -y

      - name: Checkout
        uses: actions/checkout@v2

      - name: Setup node
        uses: actions/setup-node@v2
        with:
          node-version: '16'

      - name: Setup yarn
        run: npm install -g yarn

      - name: Setup cache
        uses: actions/cache@v2
        id: yarn-cache
        with:
          path: '**/node_modules'
          key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}

      - name: Install dependencies
        if: steps.yarn-cache.outputs.cache-hit != 'true'
        run: yarn

      - name: Lint
        run: yarn lint
