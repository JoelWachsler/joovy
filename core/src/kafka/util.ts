import { Kafka } from 'kafkajs'
import { defer, from, mapTo, mergeMap, Observable, tap } from 'rxjs'
import initConfig from '../config'
import logger from '../logger'
import { KMessage } from './kafka'

let kafka: Kafka

/**
 * Internal implementation detail, do not use.
 */
export const createKafkaClient = () => {
  if (!kafka) {
    const kafkaConfig = initConfig().kafka

    kafka = new Kafka({
      clientId: kafkaConfig.clientId,
      brokers: kafkaConfig.brokers,
    })
  }

  return kafka
}

type CreateConsumerArgs = {
  topic: string
  groupId: string
}

export const createConsumer = <T extends KMessage = KMessage>(args: CreateConsumerArgs) => {
  const { groupId, topic } = args

  const create = defer(() => {
    logger.info(`Creating consumer with the following args: ${JSON.stringify(args)}`)
    const consumer = createKafkaClient().consumer({ groupId })

    return from(consumer.connect()).pipe(mapTo(consumer))
  })

  const connectToTopic = create.pipe(mergeMap(consumer => {
    logger.info(`Subscribing to topic: ${topic}...`)
    return from(consumer.subscribe({ topic, fromBeginning: true })).pipe(
      tap(() => logger.info(`Subscribed to topic: ${topic}!`)),
      mapTo(consumer),
    )
  }))

  return connectToTopic.pipe(mergeMap(consumer => {
    return new Observable<T>(subscribe => {
      consumer.run({
        async eachMessage({ message, topic }) {
          const msgAsString = message.value?.toString()
          if (!msgAsString) {
            logger.error(`Invalid message: ${message} for topic: ${topic}`)
            return
          }

          logger.info(`New incoming message: ${JSON.stringify(msgAsString)} on topic: ${topic}`)

          subscribe.next(JSON.parse(msgAsString))
        },
      })
    })
  }))
}