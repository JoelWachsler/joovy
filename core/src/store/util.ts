import { defer } from 'rxjs'
import JEvent from '../jevent/JEvent'

const messageStore = new Map<string, Map<string, unknown>>()

export const getOrCreateObjectStore = (id: string) => {
  let store = messageStore.get(id)
  if (!store) {
    store = new Map()
    messageStore.set(id, store)
  }

  return store
}

export const removeObjectStore = (event: JEvent) => {
  return defer(() => {
    const channelId = event.message.channelId
    messageStore.delete(channelId)

    return event.result({ store: { event: 'removed', id: channelId } })
  })
}
