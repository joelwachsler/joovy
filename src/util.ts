
export namespace Util {

  const apply = <T>(t: T) => {
  }
}
