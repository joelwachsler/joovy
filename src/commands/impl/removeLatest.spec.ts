import { createTestEvent, handle, hot } from '../../test/util'

test('should skip the last added track', () => {
  const play = createTestEvent({ content: '/play test' })
  const play2 = createTestEvent({ content: '/play test2' })
  const removeLatest = createTestEvent({ content: '/removelatest' })

  const messages = handle(hot('abc', { a: play, b: play2, c: removeLatest }))
  expect(messages).toMatchSnapshot()
})

test('should skip the current playing track if that\'s the last one added', () => {
  const play = createTestEvent({ content: '/play test' })
  const removeLatest = createTestEvent({ content: '/removelatest' })

  const messages = handle(hot('ab', { a: play, b: removeLatest }))
  expect(messages).toMatchSnapshot()
})

test('should stop the player if the user has removed the last track and skipped the current one', () => {
  const play = createTestEvent({ content: '/play test' })
  const play2 = createTestEvent({ content: '/play test2' })
  const removeLatest = createTestEvent({ content: '/removelatest' })
  const skip = createTestEvent({ content: '/skip' })

  const messages = handle(hot('abcd', { a: play, b: play2, c: removeLatest, d: skip }))
  expect(messages).toMatchSnapshot()
})

test('should increment current track correctly when tracks are skipped', () => {
  const play = createTestEvent({ content: '/play test' })
  const play2 = createTestEvent({ content: '/play test2' })
  const removeLatest = createTestEvent({ content: '/removelatest' })
  const skip = createTestEvent({ content: '/skip' })
  const printQueue = createTestEvent({ content: '/queue' })

  const messages = handle(hot('abcdae', {
    a: play,
    b: play2,
    c: removeLatest,
    d: skip,
    e: printQueue,
  }))
  expect(messages).toMatchSnapshot()
})
