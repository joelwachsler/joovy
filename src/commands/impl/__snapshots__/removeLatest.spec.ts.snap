// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should increment current track correctly when tracks are skipped 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/removelatest",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been removed.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/skip",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "track": "was removed, skipping",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "skipped": true,
      },
    },
  },
  TestMessageValue {
    "frame": 4,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 4,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 4,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 4,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/queue",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": \\"Queue\\",
  \\"type\\": \\"rich\\",
  \\"description\\": \\"\`0)\` [test (1:10)](https://test.com) [<@testAuthorId>]\\\\n\`1)\` ~~[test2 (1:10)](https://test2.com) [<@testAuthorId>]~~\\\\n⬐ current track\\\\n\`2)\` [test (1:10)](https://test.com) [<@testAuthorId>]\\\\n⬑ current track\\",
  \\"url\\": null,
  \\"timestamp\\": \\"2000-02-01T00:00:00.000Z\\",
  \\"color\\": 16758815,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 9,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 9,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;

exports[`should skip the current playing track if that's the last one added 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/removelatest",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been removed.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;

exports[`should skip the last added track 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/removelatest",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been removed.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "track": "was removed, skipping",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;

exports[`should stop the player if the user has removed the last track and skipped the current one 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/removelatest",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been removed.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/skip",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "track": "was removed, skipping",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "skipped": true,
      },
    },
  },
]
`;
